aten/src/ATen/native/Unique.cpp

```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+  mask_acc[0] = true;
+
+  // we can parallel on [1, numel) but we need to make sure it has
+  // the same parallel scope with the next loop
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    if (begin == 0) { begin += 1; }
+    for (const auto i : c10::irange(begin, end)) {
+      mask_acc[i] = input_sorted_data[i] != input_sorted_data[i - 1];
+    }
+  });
+
+  // handle equal the NaN's
+  bool last_element_isnan = _isnan<scalar_t>(input_sorted_data[numel - 1]);
+  if (input.is_floating_point() && equal_nan && last_element_isnan) {
+    int64_t firstnan_index = numel;
+    while (_isnan<scalar_t>(input_sorted_data[firstnan_index - 1])) {
+      firstnan_index--;
+    }
+    mask_acc[firstnan_index] = true;
+    if (firstnan_index + 1 < numel) {
+      for (const auto i : c10::irange(firstnan_index + 1, numel)) {
+        mask_acc[i] = false;
+      }
+    }
+  }
+
+  // calculate unique count from each thread
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
+    for (const auto i : c10::irange(begin, end)) {
+      if (mask_acc[i]) {
+        unique_count_thread[tid]++;
+      }
+    }
+  });
+
+  // calculate total count of unique and thread offset in output
+  int64_t unique_count = std::accumulate(unique_count_thread.begin(), unique_count_thread.end(), 0);
+  int64_t sum = 0;
+  for (const auto t : c10::irange(num_threads)) {
+    offset_thread[t] = sum;
+    sum += unique_count_thread[t];
+  }
+
+  output.resize_({unique_count});
+  scalar_t* output_data = output.data_ptr<scalar_t>();
+
+  int64_t* inverse_indices_data = nullptr;
+  if (return_inverse) {
+    inverse_indices.resize_(input.sizes());
+    inverse_indices_data = inverse_indices.data_ptr<int64_t>();
+  }
+
+  int64_t* counts_data = nullptr;
+  int64_t* unique_index_data = nullptr;
+  if (return_counts) {
+    counts.resize_({unique_count});
+    counts_data = counts.data_ptr<int64_t>();
+
+    unique_index.resize_({unique_count + 1});
+    unique_index_data = unique_index.data_ptr<int64_t>();
+    unique_index_data[unique_count] = numel;
+  }
+
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
+    int64_t offset = offset_thread[tid];
+
+    for (const auto i : c10::irange(begin, end)) {
+      if (mask_acc[i]) {
+        output_data[offset] = input_sorted_data[i];
+        if (return_counts) {
+          unique_index_data[offset] = i;
+        }
+        offset++;
+      }
+
+      if (return_inverse) {
+        int64_t inverse_index = offset - 1;
+        int64_t perm = indices_data[i];
+        inverse_indices_data[perm] = inverse_index;
+      }
+    }
+  });
+
+  if (return_counts) {
+    // do diff to get count
+    at::parallel_for(0, unique_count, 0, [&](int64_t begin, int64_t end) {
+      for (const auto i : c10::irange(begin, end)) {
+        counts_data[i] = unique_index_data[i + 1] - unique_index_data[i];
+      }
+    });
```

Comment:
This looks like `unique_index[1:] - unique_index[:-1]`? So it appears to be just a difference of slices and we can use high level functions here.

Reply:
Yes, we can do a `torch.diff` here. But like the previously one, performance would be a little bit slower, roughly add another ~5% overhead.